src/main/java/com/commercetools/sync/types/utils/TypeUpdateFieldDefinitionActionUtils.java

```diff
@@ -0,0 +1,277 @@
+     *
+     * <p>Note: If the field type field changes, the old field definition is removed and the new field
+     *     definition is added with the new field type.
+     *
+     * @param oldFieldDefinitions       the list of old {@link FieldDefinition}s.
+     * @param newFieldDefinitions       the list of new {@link FieldDefinition}s.
+     * @return a list of field definition update actions if there are field that are not existing
+     *         in the new draft. If the field definition still exists in the new draft, then compare the field
+     *         definition fields (name, label, etc..), and add the computed actions to the list of update actions.
+     *         Otherwise, if the field definitions are identical, an empty optional is returned.
+     */
+    @Nonnull
+    private static List<UpdateAction<Type>> buildRemoveFieldDefinitionOrFieldDefinitionUpdateActions(
+            @Nonnull final List<FieldDefinition> oldFieldDefinitions,
+            @Nonnull final List<FieldDefinition> newFieldDefinitions) {
+
+        final Map<String, FieldDefinition> newFieldDefinitionsNameMap =
+                newFieldDefinitions
+                        .stream().collect(
+                        toMap(FieldDefinition::getName, fieldDefinition -> fieldDefinition,
+                                (fieldDefinitionA, fieldDefinitionB) -> {
+                                    throw new DuplicateNameException(format("Field definitions have duplicated names. "
+                                                    + "Duplicated field definition name: '%s'. "
+                                                    + "Field definitions names are expected "
+                                                    + "to be unique inside their type.",
+                                            fieldDefinitionA.getName()));
+                                }
+                        ));
+
+        return oldFieldDefinitions
+                .stream()
+                .map(oldFieldDefinition -> {
+                    final String oldFieldDefinitionName = oldFieldDefinition.getName();
+                    final FieldDefinition matchingNewFieldDefinition =
+                            newFieldDefinitionsNameMap.get(oldFieldDefinitionName);
+
+                    if (matchingNewFieldDefinition == null) {
+                        return singletonList(RemoveFieldDefinition.of(oldFieldDefinitionName));
+                    } else {
+                        if (matchingNewFieldDefinition.getType() != null) {
+                            // field type is required so if null we let commercetools to throw exception
+                            if (haveSameFieldType(oldFieldDefinition, matchingNewFieldDefinition)) {
+                                return buildActions(oldFieldDefinition, matchingNewFieldDefinition);
+                            } else {
+                                return Arrays.asList(
```

Comment:
I would add here as a comment that this is a work around for changing the type of the definition. Since there is not action, so we remove then add again.

Because it might not be clear without a comment why we do this.

Also for the product type sync pls :)

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/304/commits/18970e91045e6ba4cc5756013fb24b5c1e9ecb34